core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -386,7 +387,7 @@ public final void runInContext() {
             }
           } catch (Throwable t) {
             stream.cancel(Status.CANCELLED.withCause(t).withDescription("Failed to read message."));
-            return;
+            Throwables.propagateIfPossible(t);
```

Comment:
So, just to be clear (if it wasn't/if it matters), the exception (an `Error`, actually ...) is being thrown in the `onNext` client callback of an async call, i.e., it's a `StreamObserver` passed as part of the async call to receive the result. The result from the server was fine; it was the client processing code that was bolluxed up.

And when I said the "sent on the stream", I did mean as part of the cancel payload. But I'm still pretty weak on how the streams are implemented.

I do get the subtitles of this/that it's not necessarily an easy fix. It's fundamentally impossible for a framework to unambiguously attribute a failure to a client or a server. It's too dependent on the code.

Re: the `onError` comment, are you thinking that an exception thrown in an `onNext` handler causes a call to `onError`? So that starts to conflate/combine local exceptions with server/transport-generated errors? Hrm ...

And I still have concerns about any processing/delay in notification of `Error`s which can be pretty severe ...